mui/native_plugin.py

```diff
@@ -0,0 +1,69 @@
+from typing import Callable
+from manticore.core.plugin import Plugin
+from mui.hook_manager import NativeHookManager
+
+
+class RebaseHooksPlugin(Plugin):
+    def __init__(self, mgr: NativeHookManager, find_f: Callable, avoid_f: Callable):
+        super().__init__()
+        self.find_f = find_f
+        self.avoid_f = avoid_f
+        self.filename = mgr.bv.file.original_filename
+        self.base = 0
+        self.loaded = False
+
+        # Hooks
+        self.find = mgr.list_find_hooks()
+        self.avoid = mgr.list_avoid_hooks()
+        self.custom_hooks = [(addr, func) for addr, func in mgr.list_custom_hooks().items()]
+        self.global_hooks = list(mgr.list_global_hooks().values())
+
+        # Close binary view, don't use after this
+        mgr.bv.file.close()
+
+    def on_register(self):
+        """Called by parent manticore on registration (enable global hooks)"""
+        m = self.manticore
+        for func in self.global_hooks:
+            exec(func, {"bv": None, "m": m})
+
+        # Hack: call m.add_hook once to subscribe _hook_callback
+        m.hook(0)(lambda state: state)
+
+    def did_map_memory_callback(self, _state, addr, _size, _perms, filename, offset, _addr):
+        """Rebases hooks from library to loaded base address"""
+        # If binary base is being loaded
+        if filename == self.filename and offset == 0:
+            print(f"{filename} mapped @ {addr:#x}")
+            with self.locked_context():
+                m = self.manticore
+                self.loaded = True
+                self.base = addr
+
+                # Hack: make m.subscribe != None to prevent error in mcore
+                # We don't need subscribe, we just want our hooks to be in m._hooks
+                m.subscribe = lambda x, y: None
```

Comment:
This doesn't seem safe. I don't quite understand what `m.subscribe` has to do with adding more hooks (I haven't looked at that code in a while), but we should at least first check that `m.subscribe == None` and then save whatever `m.subscribe` was into a temporary variable that is restored at the end.

Reply:
I read the explanation below and I still think we should practice defensive programming to either check that `m.subscribe == None` before doing the reassignment or save the original value of `m.subscribe` into a temporary variable and restore the original value at the end.